### ISSUE TITLE
Self-shadow pop-in

### ISSUE THREAD

**Issue by star309** [elishacloud/Silent-Hill-2-Enhancements#269]
Self-shadows do not properly resolve, but instead just pop in. For example, in the apartment complex at the beginning of the game, when you slowly walk into the light cone produced by a lighting fixture overhead, shadows just unnaturally pop in, instead of slowly resolving. At the edges of the light cone, the shadows should be fuzzy and then as the intensity of the light increases, the shadows should become more defined, or in other words, less fuzzy. The shadows look fuzzy at lower intensities of light, because interference patterns created by light waves are more apparent at lower light intensities. This is how light naturally behaves. Shadows do not suddenly pop into view in real life, instead they gradually resolve in direct proportion to the light intensity in any given volume of space. 

Basically, what I am saying is that shadows shouldn't just suddenly appear when you approach a light source, they should gradually resolve from a very fuzzy mess into clear outlines as you get closer and closer to the light source. This is because light behaves like a wave. 

So, if you could just assign incremental numeric luminosity levels to static light sources which typically produce self-shadows, and then have the resolution of the shadows increase with each increment, then the shadows could naturally resolve instead of just popping in.

**Comment by Polymega** [comment]
This is a limitation of the game engine as when used during SH2's development and the observations you've noted are correct, normal behavior across all versions of the game. (Not including HD Collection.)

You're totally right that the shadow behavior is not real world accurate, and in many areas the shadows can appear to wildly pop in and out, depending on your movement. 

The HD Collection tried to improve the shadows  through custom shaders and, while the effort was commendable, it impacted shadow behavior for the worse elsewhere throughout the game. 

Additionally, their attempt didn't fade shadows depending on proximity to a light source; they just tried to fix other aspects with the shadows.

- https://twitter.com/theRatiocinator/status/1186839931399954432?s=09
- https://twitter.com/theRatiocinator/status/1155620160771940352?s=09
- https://twitter.com/theRatiocinator/status/1155620811639853056?s=09

> So, if you could just assign incremental numeric luminosity levels to static light sources which typically produce self-shadows, and then have the resolution of the shadows increase with each increment, then the shadows could naturally resolve instead of just popping in.

This would be very nice to do, and would add a gorgeous coat of paint to the shadows, but I'm not sure if that's possible for us to do at this time. 

The specialized light sources in the game, as far as we understand, only use a simple on/off byte to signify whenever James is in or out of the light source. There's no way to incrementally change it if there's no code already in the game to monitor such values for us to do so.

**Comment by star309** [comment]
That is very unfortunate. I guess, you'd have to modify the engine in order to accomplish that. But on the flip side, that would produce high-quality volumetric shadows, especially if each light source had at least a few hundred luminosity levels, then diffuse light, at the edges of the light cone, would produce fuzzy shadows, whilst focused (intense) light, in the centre of the light cone, would produce deep, clean shadows. And of course, in that case if there were two light sources, then their intensity grids would overlap, and these two light sources would produce realistic double shadows. And shadows, in general, would just gradually resolve as you drew closer to a light source. 

I used to work on such a volumetric shadow implementation for the original Watch Dogs, but this implementation was later scraped in favour of a simplified version, because next-gen consoles (at that time) had potato CPUs that could not possibly handle such shadows, to say nothing of incredibly unwieldy occlusion maps. Basically, even diffuse light sources will produce well-defined shadows if you stand really close to a surface which you occlude. For example, if you draw closer to a brick wall your self-shadow should become more defined. That meant that we had to build invisible occlusion maps for all surfaces which made shadows more defined. Our solution (Volumetric Shadow Mapping) made shadows look insanely good, if light sources had at least a few hundred luminosity levels, and extra junky if light sources had about a dozen luminosity levels, which is all that the PS4 and XBOX ONE's anemic CPUs could handle. I mean, we were, basically, relying very heavily on the CPU to render the shadows. 

By the way, you can _really_ see volumetric (double) shadows and occlusion maps at work in this comparison: https://www.youtube.com/watch?v=sX4HTUuCy3k

**Comment by Polymega** [comment]
That E3 footage was mind-blowing for Watch Dogs, from a technical standpoint. Still is, without question, to this day. Great work on the game and sorry to hear of the compromises you had to make, as I'm sure it was hard to slice away so many of the cool things your lighting team accomplished. I still thought the game looked plenty fine on release, and can only be a little saddened over the thought of any backlash you guys received over it (reminds me of the Spider-Man puddle fiasco). Anyway, I'm sure you understand the limitations we face then as well, especially as we have no source code to work from.

**Comment by star309** [comment]
Visual compromises were not the biggest problem with Watch Dogs. It's the story and the game's finale that suffered the most from potato CPUs in the PS4 and XBOX ONE. The game was supposed to end with a massive shoot-out at a huge glass building outside the city. You were supposed to prevent a mass shooting at the headquarters of this tech company, but failed. As you entered the glass building you'd see massive streams of people rushing by, people bumping into each other, some falling, and you could actually help those who fell down. Then, as you started shooting, the glass walls and fixtures would start breaking into tens of thousands of glistening shards. The varicolored reflective marble floor on the ground level would become littered with shards of glass and other wreckage. Everything was supposed to be physically simulated, from reflective, slowly growing pools of deep-red blood around dead bodies to realistically breaking glass, to long soft shadows cast by the setting sun outside the glass walls. And as you were walking up the winding glass-walled staircase to the second floor, you'd see trees, grass and underbrush swaying in the wind as police cars, ambulances and news crews were starting to arrive at the scene, you'd hear police yell through megaphones and see EMTs giving out blankets to thousands of survivors gathered outside behind the police lines, you were supposed to see a violent reaction to this event by a living city as you could still hear shots ring out in the background. Eventually, as you progressed further and further, the west wing would catch on fire as the sun set, and everything would be illuminated by the flames, which would produce realistic dancing shadows on the filing cabinets, the floors and other fixtures while the volumetric physics-based smoke would slowly fill all the rooms and stream in twisting gusts outside through the holes in the walls. That was supposed to be the grand next-gen finale for Watch Dogs, unfortunately it was technically impossible at the time. As a matter of fact, the whole city was supposed to be governed by a sophisticated AI, and the game world was not meant to disappear as you entered a building. Of course, in the final version, the physics simulations had to be pared down and the AI had to be lobotomised, we even had to pare down the number of people and cars on the streets. The game was, originally, built to run on the 10-core, 20-thread CELL-2 processor, which, Sony told us in 2011, they wanted to put in the PS4. Unfortunately, things did not work out like that. But new Ryzen chips in the PS5 and the XBOX Series X should be powerful enough to enable such experiences, which were never possible on current-gen consoles or even on quad-core Intel CPUs.
